crates/span/src/ast_id.rs

```diff
@@ -152,126 +548,345 @@ impl AstIdMap {
     pub fn from_source(node: &SyntaxNode) -> AstIdMap {
         assert!(node.parent().is_none());
         let mut res = AstIdMap::default();
+        let mut index_map = ErasedAstIdNextIndexMap::default();
+
+        // Ensure we allocate the root.
+        res.arena.alloc((SyntaxNodePtr::new(node), ROOT_ERASED_FILE_AST_ID));
 
-        // make sure to allocate the root node
-        if !should_alloc_id(node.kind()) {
-            res.alloc(node);
-        }
         // By walking the tree in breadth-first order we make sure that parents
         // get lower ids then children. That is, adding a new child does not
         // change parent's id. This means that, say, adding a new function to a
         // trait does not change ids of top-level items, which helps caching.
-        bdfs(node, |it| {
-            if should_alloc_id(it.kind()) {
-                res.alloc(&it);
-                TreeOrder::BreadthFirst
-            } else {
-                TreeOrder::DepthFirst
+
+        // This contains the stack of the `BlockExpr`s we are under. We do this
+        // so we only allocate `BlockExpr`s if they contain items.
+        // The general idea is: when we enter a block we push `(block, false)` here.
+        // Items inside the block are attributed to the block's container, not the block.
+        // For the first item we find inside a block, we make this `(block, true)`
+        // and create an ast id for the block. When exiting the block we pop it,
+        // whether or not we created an ast id for it.
+        // It may seem that with this setup we will generate an ID for blocks that
+        // have no items directly but have items inside other items inside them.
+        // This is true, but it doesn't matter, because such blocks can't exist.
+        // After all, the block will then contain the *outer* item, so we allocate
+        // an ID for it anyway.
+        let mut blocks = Vec::new();
+        let mut curr_layer = vec![(node.clone(), None)];
```

Comment:
can we undo the inlining of `bdfs`? Imo this hurts readability of the code quite a bit

Reply:
I prefer not because with the handling for block exprs it's tightly coupled with the rest of the code.